url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/Nullable.java

```diff
@@ -0,0 +1,163 @@
+package fri.patterns.interpreter.parsergenerator.parsertables;
+
+import java.util.*;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+
+/**
+	Nullability of a nonterminal means that it can be "nothing", i.e.
+	there is a rule that contains the nonterminal on the left side and
+	no symbol on the right side (empty right side), expressing an optional
+	rule.
+	<p>
+	<b>Algorithm:</b><br>
+	Sort rules by size.
+	Search every rule that contain the nonterminal on left side.
+	Scan the left side of every rule until a terminal or a non-nullable
+	nonterminal appears. If found, rule is not nullable, else it is nullable.
+	If one of the rules that derive nonterminal is nullable, this nonterminal
+	is nullable.
+
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+class Nullable extends Hashtable
+{
+	/** The special empty symbol. */
+	public static final String NULL = "";
+	
+	/**
+		Explores the nullability of all nonterminals in syntax. Provides nullability
+		as Boolean within this Map, key is nonterminal.
+		@param syntax the grammar to scan for nonterminals and their nullability.
+		@param nonterminals pre-built list of nonterminals.
+	*/
+	public Nullable(Syntax syntax, List nonterminals)
+		throws ParserBuildException
+	{
+		// loop nonterminals
+		Map done = new Hashtable(nonterminals.size());	// avoid endless loops
+		for (int i = 0; i < nonterminals.size(); i++)	{
+			String nt = (String) nonterminals.get(i);
+			checkNullability(syntax, nt, done);
+		}
+	}
+
+
+	/** Returns true if passed nonterminal is nullable, else false. */
+	public boolean isNullable(String nonterminal)	{
+		Boolean nullable = (Boolean) get(nonterminal);
+		return nullable.booleanValue();
+	}
+
+
+	private boolean checkNullability(Syntax syntax, String nonterm, Map done)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 67-73 into a separate method.